src/callbacks_stage/positivity_shallow_water_dg1d.jl

```diff
@@ -84,4 +86,89 @@ function limiter_shallow_water!(u, threshold::Real, variable,
 
     return nothing
 end
+
+# !!! warning "Experimental code"
+#     This is an experimental feature and may change in future releases.
+
+# Note that for the `ShallowWaterMultiLayerEquations1D` only the waterheight `h` is limited in
+# each layer. Furthermore, a velocity desingularization is applied after the limiting to avoid
+# numerical problems near dry states.
+function limiter_shallow_water!(u, threshold::Real, variable,
+                                mesh::Trixi.AbstractMesh{1},
+                                equations::ShallowWaterMultiLayerEquations1D,
+                                dg::DGSEM, cache)
+    @unpack weights = dg.basis
+
+    Trixi.@threaded for element in eachelement(dg, cache)
+        # Limit layerwise
+        for m in eachlayer(equations)
+            # determine minimum value
+            value_min = typemax(eltype(u))
+            for i in eachnode(dg)
+                u_node = get_node_vars(u, equations, dg, i, element)
+                value_min = min(value_min, variable(u_node, equations)[m])
+            end
+
+            # detect if limiting is necessary
+            value_min < threshold - eps() || continue
+
+            # compute mean value
+            u_mean = zero(get_node_vars(u, equations, dg, 1, element))
+            for i in eachnode(dg)
+                u_node = get_node_vars(u, equations, dg, i, element)
+                u_mean += u_node * weights[i]
+            end
+
+            # note that the reference element is [-1,1]^ndims(dg), thus the weights sum to 2
+            u_mean = u_mean / 2^ndims(mesh)
+
+            # We compute the value directly with the mean values.
+            # The waterheight `h` is limited independently in each layer.
+            value_mean = variable(u_mean, equations)[m]
+            theta = (value_mean - threshold) / (value_mean - value_min)
+
+            for i in eachnode(dg)
+                u_node = get_node_vars(u, equations, dg, i, element)
+                h_node = waterheight(u_node, equations)[m]
+                h_mean = waterheight(u_mean, equations)[m]
+
+                u[m, i, element] = theta * h_node + (1 - theta) * h_mean
+            end
+        end
+    end
+
+    # "Safety" application of the wet/dry thresholds over all the DG nodes
+    # on the current `element` after the limiting above in order to avoid dry nodes.
+    # If the value_mean < threshold before applying limiter, there
+    # could still be dry nodes afterwards due to logic of the limiting.
+    # Additionally, a velocity desingularization is applied to avoid numerical 
+    # problems near dry states.
+    Trixi.@threaded for element in eachelement(dg, cache)
+        for i in eachnode(dg)
+            u_node = get_node_vars(u, equations, dg, i, element)
+
+            h = MVector(waterheight(u_node, equations))
+            hv = MVector(momentum(u_node, equations))
+            b = u_node[end]
+
+            # Velocity desingularization
```

Comment:
I would add a reference to once of the Chertock papers where this desingularization procedure comes from and is motivated.

Reply:
I have added a reference to the docstring of [PositivityPreservingLimiterShallowWater](https://github.com/trixi-framework/TrixiShallowWater.jl/pull/38/commits/3d6d89a9381cd544f4ee470550c38ee47df42bf1#diff-974baa15d0336babce616da61ced079a309be41eb301b19f0b31fddcce888fc5L39) in b9302cd28ae88a0a597f1fad0e28ddad100db9d4.
I believe this is a good place, since this is also where the desingularization is implemented.